---
title: "RNA-seq workflow: annotation of RNA-seq results"
subtitle: CSAMA 2018 version 
author:
- name: Johannes Rainer
  affiliation: Institute for Biomedicine, Eurac Research, Italy
- name: Michael I. Love
  affiliation: Departments of Biostatistics and Genetics, UNC-Chapel Hill, Chapel Hill, NC, US
- name: Charlotte Soneson
  affiliation: Institute for Molecular Life Sciences, Univ. Zurich, CH
date: 12 July 2018
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Annotation of RNA-seq results}
  %\VignetteEngine{knitr::rmarkdown}
---

<!-- to compile this: rmarkdown::render("annotation-with-ensembldb.Rmd") -->

<!--
     # a list of all required libraries:
     reqlibs = sub(".*library\\(\"(.*?)\"\\).*","\\1",grep("library\\(",readLines("annotation-with-ensembldb.Rmd"),value=TRUE))
     find.package(reqlibs)
-->


```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("BiocStyle")
library("knitr")
library("rmarkdown")
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE,
               cache = FALSE, fig.width = 5, fig.height = 5)
library(AnnotationHub)
library(ensembldb)
```

# Abstract

This document describes annotation of RNA-seq results using resources from
Bioconductor, specifically the `r Biocpkg("ensembldb")` package.

# Introduction

One of the central steps in the analysis of genomic data is the annotation of
the quantified entities to biologically more relevant representations such as
transcripts, genes or proteins. Such annotations enable for example pathway
enrichment analyses and ease the interpretation of the results. Bioconductor
provides a large variety of annotation packages and resources, most of them
supporting the `r Biocpkg("AnnotationDbi")` interface which enables a unified
way to retrieve annotations for given identifiers. The annotation resources from
Bioconductor range from web-based tools such as `r Biocpkg("biomaRt")` to
packages working with pre-build databases containing all identifier mappings for
a certain species (the `*.org` packages such as `org.Hs.eg.db`) or providing
gene and transcript models, such as databases build by `r
Biocpkg("GenomicFeatures")` (`TxDb` databases/packages) and `r
Biocpkg("ensembldb")` (`EnsDb` databases/packages).

In this workflow we use the `r Biocpkg("ensembldb")` package to annotate the
result tables from Tuesday's end-to-end RNAseq lab. Pre-build `EnsDb` databases
for all core species and a range of Ensembl releases are available at the `r
Biocpkg("AnnotationHub")` resource. These `EnsDb` databases contain, along with
genomic coordinates of exons, transcripts and genes, also mappings to NCBI
identifiers and protein annotations, including their amino acid sequences and
positions of protein domains within these.


# Annotation of RNA-seq results to gene identifiers

We first load the result table from Tuesday's RNAseq lab and inspect the content
of this table. The task in this section is then to annotate the quantified
entities in this table to other identifiers.

```{r load-table}
top_table <- read.csv("data/res_lfc-1_FDR-5.csv", row.names = 1)
head(top_table)
```

In this result table rows are identified by Ensembl gene IDs. The Salmon
quantification used in the analysis leading to this table was based on Gencode
v27 transcripts which is linked to Ensembl release 91. We thus load below an
`EnsDb` database with human annotations for Ensembl release 91 from the `r
Biocpkg("AnnotationHub")` resource. The parameter `localHub = TRUE` ensures that
we list and load only locally stored resources, due to the very limited internet
connectivity at the lab site.

```{r annotationhub}
library(AnnotationHub)
library(ensembldb)
ah <- AnnotationHub(localHub = TRUE)

ah_91 <- query(ah, "EnsDb.Hsapiens.v91")
ah_91

## Load the database
edb <- ah[[names(ah_91)]]
```

If we had a working internet connection we could simply list all available
`EnsDb` databases with `query(ah, "EnsDb")` (after loading `AnnotationHub` with
`localHub = FALSE`). We can now use the `ah_91` variable to access the data in
the database. With `listColumns` function we can for example list all of the
available annotation columns in the database.

```{r listcolumns}
listColumns(edb)
```

We could get data from any of these database columns by passing the respective
column name(s) along with the `column` parameter to any function retrieving
annotations from an `EnsDb` database. Gene-related annotations can be retrieved
with the `genes` function, that by default (and similar to the `genes` function
from the `r Biocpkg("GenomicFeatures")` package) returns a `GRanges` object with
the genomic coordinates of the genes and additional annotation columns in the
`GRanges`' *metadata* columns. In our case we do however not want to fetch
annotations for all human genes, but only for the genes of our top table. For
that we will use the filtering framework from the `ensembldb`.

Before retrieving any annotations we have however to strip the version from our
Ensembl gene identifiers (e.g. the `.9` from `ENSG00000189221.9`) because
`ensembldb` uses IDs without version information.

```{r stripVersion}
rownames(top_table) <- sub("\\.[0-9]*$", "", rownames(top_table))
```

A call to `genes` would now fetch the annotations for all human genes. While we
could subset the returned result object based on our gene identifiers, it is
more elegant (and faster) to use `ensembldb`'s filtering framework to get only
annotations for our set of genes. We hence pass the filter expression `~ gene_id
== rownames(top_table)` with the `filter` parameter to the `genes`
function. Filter expressions have to be written in the form of a formula
(i.e. starting with `~`) and support any logical R expression and any database
column/field in the `EnsDb` database (use `supportedFilters(edb)` to list all
supported fields). Also we specify `return.type = "data.frame"` to tell the
`genes` function to return the results as a `data.frame` instead of the default
`GRanges` object.

```{r retrieve}
anns <- genes(edb, filter = ~ gene_id == rownames(top_table),
              return.type = "data.frame")
head(anns)
```

Along with the genomic coordinates of the genes we got also their official gene
names, description and NCBI Entrezgene identifiers. Be aware that the order of
the genes in the returned `data.frame` is **not** the same as in `top_table`. We
thus below re-order the annotations to match the order of genes in the top table
and subsequently join the two `data.frame`s.

```{r reorder-n-join}
top_table <- cbind(anns[match(rownames(top_table), anns$gene_id), ], top_table)
```

Be aware that mappings between Ensembl and NCBI identifiers is not necessarily a
1:1 mapping. In our case we got for some Ensembl gene IDs more than one NCBI
Entrezgene ID. The column `"entrezid"` in our result table is thus a `list` with
eventually more than one Entrezgene ID in one row. Below we list the rows with
the 1:n mappings between Ensembl gene IDs and NCBI Entrezgene IDs.

```{r one-to-n-mapping}
top_table[lengths(top_table$entrezid) > 1, c("gene_id", "entrezid")]
```

Exporting such a table could be problematic and we collapse therefore Entrezgene
IDs in this columns into a single character string, with multiple IDs separated
by a semicolon (`";"`).

```{r collapse-entrezid}
top_table$entrezid <- sapply(top_table$entrezid, function(z) {
    if (any(is.na(z))) z
    else paste(z, collapse = ";")
})
```

While `genes` is the more natural way to access gene-related annotations from
the `EnsDb` database, we could also use the `AnnotationDbi` functions `select`
or `mapIds` to retrieve annotations. Note that by convention `AnnotationDbi`
functions expect column names to be specified in all capital case.

```{r select}
anns_2 <- select(edb, keys = top_table$gene_id, keytype = "GENEID",
                 columns = c("GENENAME", "ENTREZID"))
head(anns_2)
```

Be aware that by default (i.e. with `multiVals = "first"`) `select` returns only
a single NCBI Entrezgene identifier for each Ensembl gene ID, even if more than
one are annotated to the Ensembl gene.

In addition to gene names and NCBI identifiers, we might also be interested to
get the identifiers of the proteins for the genes' transcripts. Below we use the
`proteins` function to retrieve protein annotations for the provided gene
identifiers.

```{r protein-annotations}
prts <- proteins(edb, filter = ~ gene_id == top_table$gene_id,
                columns = c("tx_id", "protein_id", "uniprot_id"))
head(prts)
```

We got thus for each of the genes their (protein coding) transcripts and the
identifiers of the encoded proteins (Ensembl protein IDs and Uniprot
accessions). Be aware that also the mapping between Ensembl protein identifiers
and Uniprot accessions is not 1:1.


# Query for genes/transcripts encoding a specific protein domain

For some experiments it might be interesting to search for genes, or rather
transcripts, that encode a protein with a certain protein domain. For the
present data set we could for example search for genes with proteins encoding
the ligand binding domain of nuclear hormone receptors (such as the
glucocorticoid receptor, the gene activated by treatment with the synthetic
glucocorticoid dexamethasone and hence being mainly responsible for the
transcriptional changes observed in the present dataset). Below we thus query
the database for all such genes using the protein domain ID
[PF00104](http://pfam.xfam.org/family/PF00104) from Pfam.

```{r protein-domain-query}
## Retrieve all genes with a transcript encoding a protein with a
## certain protein ID.
hrmn_rcptr <- genes(edb, filter = ~ protein_domain_id == "PF00104",
                    return.type = "data.frame")

```

The table below lists all genes with a transcript encoding the nuclear hormone
receptor ligand binding domain.

```{r nuclear-receptors, results = "asis", echo = FALSE}
knitr::kable(hrmn_rcptr[, c("gene_id", "gene_name", "description")])
```

With that information we can now evaluate whether one of these receptors is also
among our differentially expressed genes.

```{r de-hormone-receptor}
top_table[top_table$gene_id %in% hrmn_rcptr$gene_id, ]
```

Indeed, NR4A3, encoding a member of the steroid-thyroid hormone-retinoid
receptor superfamily, is 4-fold upregulated in the data set.

In addition we could also use this information to evaluate expression of all
hormone receptors, most of them being transcription factors for some of which it
was shown that they can transcriptionally regulate their own gene as part of a
negative, or in some cell types even positive, feedback loop.


# Session information

```{r  sessionInfo }
devtools::session_info() 
```


