---
title: "Lecture 11a -- Introduction to _Bioconductor_ Annotation Resources"
author: "Martin Morgan <Martin.Morgan@RoswellPark.org>"
date: "11 July 2018"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{Lecture 11a -- Introduction to Bioconductor Annotation Resources}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```

```{r setup, echo=FALSE}
suppressPackageStartupMessages({
    library(org.Hs.eg.db)
    library(TxDb.Hsapiens.UCSC.hg38.knownGene)
    library(biomaRt)
    library(KEGGREST)
    library(AnnotationHub)
    library(GenomicRanges)
    library(ExperimentHub)
    library(SummarizedExperiment)
    library(tidyverse)
})
```

# Annotation packages

- Identifier mapping, `*.org` packages, e.g., [org.Hs.eg.db][]
- Genomic features, `TxDb.*` packages, e.g.,
  [TxDb.Hsapiens.UCSC.hg38.knownGene][]
- Sequence, e.g., [BSgenome.Hsapiens.UCSC.hg38][]

## The `select()` interface

`columns()`, `mapIds()`, `select()`

```{r, message=FALSE}
library(tidyverse)
library(org.Hs.eg.db)
columns(org.Hs.eg.db)
brca <- c("BRCA1", "BRCA2")
mapIds(org.Hs.eg.db, brca, "ENSEMBL", keytype="SYMBOL")
AnnotationDbi::select(
    org.Hs.eg.db, brca, c("ENSEMBL", "GO"), keytype = "SYMBOL"
) %>% as_tibble()

library(TxDb.Hsapiens.UCSC.hg38.knownGene)
columns(TxDb.Hsapiens.UCSC.hg38.knownGene)

library(GO.db)
columns(GO.db)
```

## Genomic features

`exons()`, `genes()`, `transcripts()`, `promoters()`, ..., `exonsBy()`, ...

```{r}
ex <- exons(TxDb.Hsapiens.UCSC.hg38.knownGene)
ex
keepStandardChromosomes(ex, pruning.mode = "coarse")
cds <- cdsBy(TxDb.Hsapiens.UCSC.hg38.knownGene, "tx", use.names=TRUE)
```

`getSeq()`, `extractTranscriptSeqs()`, ...

```{r, message=FALSE}
library(GenomicFeatures)
library(BSgenome.Hsapiens.UCSC.hg38)
rname <- "hg38 cds"
if (!NROW(BiocFileCache::bfcquery(query=rname, field="rname"))) {
    seq_cds <- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg38, cds)
    saveRDS(seq_cds, BiocFileCache::bfcnew(rname=rname))
}
seq_cds <- readRDS(BiocFileCache::bfcrpath(rname = rname))

## 'predicates' to clean cds sequences
starts <- "ATG"
stops <- c("TAA", "TAG", "TGA")
p1 <- nchar(seq_cds) %% 3 == 0
p2 <- narrow(seq_cds, 1, 3) %in% starts
p3 <- narrow(seq_cds, nchar(seq_cds) - 2, nchar(seq_cds)) %in% stops
seq_cds_clean <- seq_cds[p1 & p2 & p3]
seq_cds_clean

translate(seq_cds_clean)
```

[org.Hs.eg.db]: https://bioconductor.org/packages/org.Hs.eg.db
[TxDb.Hsapiens.UCSC.hg38.knownGene]: https://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg38.knownGene
[BSgenome.Hsapiens.UCSC.hg38]: https://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg38

## Integrated, e.g., [Organism.dplyr][]

```{r, message = FALSE}
library(Organism.dplyr)
hsap <- src_organism("TxDb.Hsapiens.UCSC.hg38.knownGene")
hsap
tbl(hsap, "id")
tbl(hsap, "ranges_tx")
left_join(tbl(hsap, "id"), tbl(hsap, "ranges_tx"))
```

[Organism.dplyr]: https://bioconductor.org/packages/Organism.dplyr

# Web resources

- [Biomart][Biomart-web] / [biomaRt][]
    - [Discover and select][biomaRt-mart] a mart and dataset
    - E.g., [annotate entrez ids with GO ids][biomaRt-map]
- [KEGG][] / [KEGGREST][]
    - Essential operations outlined in the [vignette][KEGGREST-vignette]
- ...
- 'current' but more difficult to version, less reliable, not robust
  to big queries.

```{r, message = FALSE}
library(biomaRt)

## Discover and then selected mart
ensembl <- useMart("ensembl", dataset="hsapiens_gene_ensembl")

## Gene symbols associated with GO-annotated MAP kinase
## activity (GO id GO:0004704)
getBM(attributes = c('entrezgene','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0004707',
      mart = ensembl)
```
[Biomart-web]: http://www.ensembl.org/biomart
[biomaRt]: https://bioconductor.org/packages/biomaRt
[biomaRt-mart]: http://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html#selecting-a-biomart-database-and-dataset
[biomaRt-map]: http://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html#annotate-a-set-of-entrezgene-identifiers-with-go-annotation

[KEGG]: http://www.genome.jp/kegg/pathway.html
[KEGGREST]: https://bioconductor.org/packages/KEGGREST
[KEGGREST-vignette]: http://bioconductor.org/packages/release/bioc/vignettes/KEGGREST/inst/doc/KEGGREST-vignette.html

# Annotation and Experiment Hubs

[AnnotationHub][] -- Light curation of third party resources

```{r}
library(AnnotationHub)
hub <- AnnotationHub()
hub
query(hub, "grasp")   # see library(grasp2db)
query(hub, c("release-92", "homo"))
hub[["AH61472"]]
rtracklayer::import( hub[["AH61472"]] )
```

[ExperimentHub][] -- Heavily curated data sets

```{r}
library(ExperimentHub)
hub <- ExperimentHub()
query(hub, "TCGA")   # library(curatedTCGAData)
```

- often accessible via packages

    ```{r, message=FALSE}
    library(curatedTCGAData)
    
    ## discovery
    curatedTCGAData(diseaseCode = "*", assays = "*", dry.run = TRUE)
    curatedTCGAData(diseaseCode = "COAD", assays = "RPPA*", dry.run = TRUE)
    
    ## access
    gbm <- curatedTCGAData("GBM", "RPPA*", FALSE)
    
    ## use
    experiments(gbm)[["GBM_RPPAArray-20160128"]]
    ```

[AnnotationHub]: https://bioconductor.org/packages/AnnotationHub
[ExperimentHub]: https://bioconductor.org/packages/ExperimentHub

# End matter

## Session Info

```{r}
sessionInfo()
```

## Acknowledgements

Research reported in this tutorial was supported by the National Human
Genome Research Institute and the National Cancer Institute of the
National Institutes of Health under award numbers U41HG004059 and
U24CA180996.

This project has received funding from the European Research Council
(ERC) under the European Union's Horizon 2020 research and innovation
programme (grant agreement number 633974)
