---
title: "Introduction to Bioconductor Annotation Resources"
author: "Lori Shepherd Kern <Lori.Shepherd@roswellpark.org> <br>  Johannes Rainer <johannes.rainer@eurac.edu> <br> Martin Morgan <Martin.Morgan@RoswellPark.org> <br> James W. MacDonald"
date: 12-16 June | CSAMA 2023
output:
  ioslides_presentation:
    fig_retina: null
    css: jostyle.css
    widescreen: false
vignette: >
  %\VignetteIndexEntry{Introduction to Bioconductor Annotation Resources}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

---

<style type="text/css">

slides > slide:not(.nobackground):after {
  content: '';
}

slides > slide {
    -webkit-transition:none !important;transition:none !important;
}

.build > * {
  -webkit-transition: opacity 0.1s ease-in-out;
  -webkit-transition-delay: 0.1s;
  -moz-transition: opacity 0.1s ease-in-out 0.1s;
  -o-transition: opacity 0.1s ease-in-out 0.1s;
  transition: opacity 0.1s ease-in-out 0.1s;
}

.reveal .slides section .slideContent{
    font-size: 10pt;
}

</style>


```{r style, echo = FALSE, results = 'asis'}
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```

```{r setup, echo = FALSE, message = FALSE, warning = FALSE}
packs <- c("hugene20sttranscriptcluster.db",
           "org.Hs.eg.db",
           "TxDb.Hsapiens.UCSC.hg19.knownGene",
           "Homo.sapiens",
           "BSgenome",
           "BSgenome.Hsapiens.UCSC.hg19",
           "AnnotationHub",
           "AnnotationDbi",
           "Organism.dplyr",
           "biomaRt",
           "airway",
	   "ensembldb",
	   "GenomicFeatures")
BiocManager::install(packs, ask = FALSE)
for (p in packs)
    library(p, character.only = TRUE)
```

## What is annotation data? {.build}


```{r, out.width = "80%", echo = FALSE}
knitr::include_graphics("images/RR-sketch-03.png")
```


## What is annotation data? {.notransition}


```{r, out.width = "80%", echo = FALSE}
knitr::include_graphics("images/RR-sketch-04.png")
```

- Annotation data: usually external data that is used to *make sense* of
  measured entities.

## What do we mean by annotation?

Map a known ID to other functional or positional information

```{r, out.width = "75%", echo = FALSE}

knitr::include_graphics("images/mapping2.png")

```

- Map one type of ID to another type of ID that is more familiar to
collaborators or perhaps better used for further analysis or reporting.

## Typical annotation use cases

- get HUGO gene symbols for gene identifiers (e.g. Entrezgene, Ensembl...).
- map positions on the genome to transcripts (genes).
- map genes to biological pathways.
- assign peptide sequences to proteins.
- identify metabolites based on fragment spectra.

<br>
<br>

**Annotation**: map one type of ID to another type of ID.


## Annotation resources

- NCBI: https://www.ncbi.nlm.nih.gov/ 
- Ensembl: http://www.ensembl.org/index.html
- Uniprot: https://www.uniprot.org/
- Pubchem: https://pubchem.ncbi.nlm.nih.gov/
- Reactome: https://reactome.org/
- ...

<br>
<br>

Can be downloaded or queried through an API.


## Annotation resources in **Bioconductor** {.build}

<div>
- Annotation packages: 
  - *org.Hs.eg.db*
  - *BSgenome.Hsapiens.UCSC.hg19*
  - *TxDb.Hsapiens.UCSC.hg19.knownGene*
  - *GO.db*
  - ...
</div>
<div>
- *AnnotationHub*: query and retrieve annotations, cache them locally.
</div>
<div>
- Online resources: *biomaRt*, *KEGGREST*.
</div>
<div>
- Direct access to web resources through an API <font color="#1a81c2">(be sure to BiocFileCache!)</font>
</div>

## Annotation packages in Bioconductor {.build}

<div>
- *org.Hs.eg.db*: package with various annotations for homo sapiens.
</div>
<div>

```{r, message = FALSE, warning = FALSE}
BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
org.Hs.eg.db
```

</div>

## 

<div style="position:absolute; top:45%">
**How can we get access to the data in a Bioconductor annotation resource?**
</div>

## | **A common annotation interface:** *AnnotationDbi* {.build}

<div>
The main function is `select`:

AnnotationDbi::select(*anno*, *keys*, *columns*, *keytype*)
</div>
<div>
Where

- *anno* is the annotation package (or object)
- *keys* are the IDs that we **have**
- *columns* are the values we **want**
- *keytype* is the type of key used
</div>

## Simple Example {.build}

<div>
- *Task*: annotate gene identifiers to HUGO symbols.
</div>
<div>
- The *airway* package provides a `RangedSummarizedExperiment` with results from
  an RNA-Seq experiment.
- Genes are identified by their Ensembl IDs.
</div>

<div>

```{r, warning=FALSE}
library(airway)
data(airway)
ids <- head(rownames(airway))
ids
```

</div>


```{r, warning=FALSE}
select(org.Hs.eg.db, ids, "SYMBOL", "ENSEMBL")
```

## What annotations are provided? {.build}

<div>
- Use `columns` to list annotations available in specific annotation object.

</div>
<div>

```{r}
columns(org.Hs.eg.db)
```

</div>

<div>
- Use `keytypes` to list supported *key types*.

</div>
<div>

```{r}
keytypes(org.Hs.eg.db)
```

</div>

## Another example {.build}

- Get OMIM IDs for two genes.

```{r, warning = FALSE}
brca <- c("BRCA1", "BRCA2")
select(org.Hs.eg.db, brca, c("GENENAME", "OMIM"), "SYMBOL")

```

- We no longer have a 1:1 mapping!

## Alternative: the `mapIds` function {.build}

- Same as `select` (but for single annotations!).
- Parameter `multiVals` allows to specify how to deal with 1:many mappings.

```{r, warning = FALSE}
mapIds(org.Hs.eg.db, brca, "OMIM", "SYMBOL", multiVals = "first")

```

- OK, but where's the rest?


## Choices for `multiVals` {.build}

<div>
- `multiVals = "first"` (default): take the first element.
</div>
<div>
- `multiVals = "asNA"`: `NA` is returned for any 1:many mapping.
- `multiVals = "list"`: multiple elements are collapsed into a `list`.
- `multiVals = "CharacterList"`: results are returned as a `CharacterList`.
</div>
<div>
```{r, warning = FALSE}
mapIds(org.Hs.eg.db, brca, "OMIM", "SYMBOL", multiVals = "CharacterList")
```
</div>

## But... {.build}

- *AnnotationDbi* is a very powerful framework, but:

<div>
- For some annotations or queries different approaches might be more suitable.
- Not guaranteed that all annotation packages/objects support the
  *AnnotationDbi* framework.
</div>


## What about positional annotation? {.build}

<div>
- Annotation for positions on the genome (exons, transcripts, genes).
- Annotation for positions along protein sequences (protein domains...).
- Mostly used for gene quantification in RNA-seq data.
</div>


## | Positional annotations: `TxDb` and `EnsDb` objects {.build}

<div>
- `TxDb` (*GenomicFeatures* package) and `EnsDb` (*ensembldb* package) objects
  contain positional annotations. 
- `EnsDb` contain additional annotations such as gene symbols, Entrezgene IDs,
  GC content, gene/transcript biotypes and protein annotations.
</div>
<div>
- `TxDb` and `EnsDb` resources can be installed as packages
  (e.g. *TxDb.Hsapiens.UCSC.hg19.knownGene*, *EnsDb.Hsapiens.v86*).
- Preferred way: through *AnnotationHub*.
</div>
<div>
- **Why?** 
  - Provides annotation databases for all species, all Ensembl releases.
  - Guarantees reproducibility.
</div>


## Query *AnnotationHub* {.build}

- Search annotation resource for a specific species and release.

```{r, message = FALSE}
library(AnnotationHub)
ah <- AnnotationHub()
query(ah, c("EnsDb", "hsapiens", "102"))
```


## Fetch from *AnnotationHub* {.build}

- Download and cache this annotation.

```{r, message = FALSE}
edb <- ah[["AH89180"]]
edb
```


## Extract annotations from an `EnsDb` or `TxDb` {.build}

- Using *AnnotationDbi*:

```{r, warning = FALSE}
columns(edb)
select(edb, "ENSG00000139618",
       c("SYMBOL", "SEQNAME"), "GENEID")

```


## Extract positional annototations {.build}

- `genes`, `transcripts`,`exons` to extract their genomic coordinates.
- Result returned as `GRanges` object (with additional *metadata columns*).

```{r, warning = FALSE}
genes(edb)
```

## Query data for specific genes {.build}

- Get all transcripts for a gene with the gene symbol (name) *BRCA2*.

```{r}
transcripts(edb, filter = ~ symbol == "BRCA2")
```


## For read counting/gene quantification {.build}

- Extract exons grouped by transcript for all genes encoded on chromosome *Y*.

```{r}
exs <- exonsBy(edb, by = "tx", filter = ~ seq_name == "Y")
exs
```


## Genomic sequences {.build}

<div>
- *BSgenome* packages (e.g. *BSgenome.Hsapiens.UCSC.hg19*).

  + BSgenomeForge to create your own
  +
```{r}
head(BiocManager::available("BSgenome"),3)
```
  
</div>
<div>
- *AnnotationHub*:
</div>

```{r}
query(ah, c("TwoBitFile", "GRCh38"))
```

## Extracting transcript sequences {.build}

- *Task:* Get sequences for previously extracted transcripts.

<div>
```{r}
gn <- ah[["AH49723"]]

```
</div>
<div>
```{r}
extractTranscriptSeqs(gn, exs)
```

## Extracting genomic sequences {.build}

- Get the genomic sequence for a particular region. 
- `getSeq` takes the data object and a `GRanges` defining which (sub)sequence to
  extract.

```{r}
rng <- GRanges("X:10000-15000")
getSeq(gn, rng)
```

- See *Biostrings* package for more information on how to handle and process
  sequences.


## Online resources: *biomaRt* {.build}

<div>
- Biomart is an Ensembl service providing a web API to retrieve annotations.
- Bioconductor's *biomaRt* package allows to query Ensembl Biomart servers.
</div>
<div>

```{r, warning = FALSE}
library(biomaRt)
listMarts()
```

</div>
<div>
- Connect to a Biomart server.

```{r, warning = FALSE}
mart <- useMart("ENSEMBL_MART_ENSEMBL")
```

</div>

## *biomaRt* data sets {.build}

<div>
- And we can then check for the available data sets on a particular
  server.
</div>
<div>

```{r, warning = FALSE}
listDatasets(mart) |> head()
```


- We've got annotations for almost all species.

</div>

## *biomaRt* queries {.build}

<div>
- Connect to a particular Biomart data set.

```{r, warning = FALSE}
mart <- useMart("ENSEMBL_MART_ENSEMBL","hsapiens_gene_ensembl")
```
</div>
<div>
- Query the resource using `getBM(attributes, filters, values, mart)`.

where

- *attributes* are the things we **want** (*columns*)
- *filters* are the *types of* IDs we **have** (*keytypes*)
- *values* are the IDs we **have** (*keys*)
- *mart* is the `mart` object we set up

</div>

## *biomaRt* attributes and filters

- Use `listAttributes` and `listFilters` to get available annotations and
  supported filters.

```{r, warning = FALSE}
listAttributes(mart) |> head()
listFilters(mart) |> head()

```


## *biomaRt* query {.build}

- Example: get Ensembl gene ID and chromosome names for two genes.

```{r, warning = FALSE}
sym <- c("BRCA1", "BRCA2")
getBM(c("ensembl_gene_id", "hgnc_symbol", "chromosome_name"),
      "hgnc_symbol", sym, mart)

```

<div>
**Notes**

- Requires internet connection.
- By default uses *current* release, but `listEnsemblArchives()` would allow
  to select URLs for previous (Ensembl) releases.
</div>

## What else? {.build}

<div>
- Annotations for **chemical compounds**.
- Annotation resources: [Human Metabolome Database (HMDB)](https://hmdb.ca), 
  [PubChem](https://pubchem.ncbi.nlm.nih.gov/),
  [ChEBI](https://www.ebi.ac.uk/chebi/), 
  [MassBank](https://massbank.eu/MassBank/), 
  [MoNA](https://mona.fiehnlab.ucdavis.edu/), ...
</div> 

<div>
<div style="position:absolute; left:7%; top:40%;">
<img src=images/CompoundDb.png width=200> 
</div> 

<div style="position:absolute;left:30%; top: 47.5%;">
- Store compound annotations (mass, name, chemical formula etc).
- Fragment spectra (structural information).
</div>
</div>

## Helpful Hints


- Stick whenever possible with one annotation resource (Ensembl, NCBI, UCSC).
- Ensure you're using the same annotation version through your full analysis
  (e.g. aligning reads, gene quantification, annotation of genes).
- Prefer *AnnotationHub* over annotation packages: better control of the
  annotation version.
- For other web based resources, implement caching!

## What is caching?
<br>
<br>
<div class="centered">
<font color="#1a81c2" size="20">
BiocFileCache
</font>
<br>
<font color="grey"> Local File Management</font>
<br>
<br>
</div>
<div class="centered">
<img src=images/BFC.png width=200> 
</div> 

## Motivation

- It can be time consuming to download remote resources from the web.
<div class="centered">
<img src=images/waiting.jpg width=200> 
</div> 
- Better local file organization


##  {data-background=images/BFCfunc.png data-background-size="100%"}

## Simple example {.build}

<div>
```{r, warning = FALSE}
library(BiocFileCache)
bfc = BiocFileCache(cache=tempdir())
bfc
```
</div>

<div>
```{r, warning = FALSE}
bfcadd(
bfc,
rname="GSE52778 Series Matrix",
fpath="ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE52nnn/GSE52778/matrix/GSE52778_series_matrix.txt.gz")
```
</div>

<div>
```{r, warning = FALSE}
bfc
```
</div>

## Simple Example (cont) {.build}
 
Add a placehold to use in save/write 

<div>
```{r, warning = FALSE}
pathToSave = bfcnew(bfc, rname="MyRDSFile", ext=".rds")
pathToSave
```
</div>
<div>
```{r, warning = FALSE}
bfcinfo(bfc)
```
</div>

## Simple Example (cont) {.build}

<div>
Find files by query

```{r, warning = FALSE}
bfcquery(bfc, query="RDS")
```

Retrieve the local path

```{r, warning = FALSE}
bfcrpath(bfc, rids="BFC2")
```
</div>
<div>
There is a way to add metadata to the cache. That metadata is also
query-able. See additional slides and labs for more info on this.
</div>

## Implementations {.build}

<div>
- Directly 
- Package backend to manage package data
</div>
<div>
<br>
<br>
<div class="centered">
Two of those packages are the ExperimentHub and AnnotationHub
<br>
"The Hubs"
<br>
<img src=images/EH.png width=200> 
<img src=images/AH.png width=200> 
</div> 

## AnnotationHub/ExperimentHub "The Hubs"

<div class="centered">
What is the AnnotationHub and ExperimentHub?
<br>
<br>
Databases
<br>
That store metadata about resources,
<br>
Including download location

## AnnotationHub() / ExperimentHub()

- create a hub object
- sqlite database backend
    + metadata about the resources including download location
- Files are stored remotely and downloaded as needed/requested
    + Stored elsewhere (Zenodo, figshare, etc - preapproval recommended)
    + Bioconductor default Microsoft Azure Data Lake
- After downloaded, file is cached (with BiocFileCache) for quick access for future runs

## How do you find resources of interest?

<br>
<br>

- query()
    + general search (recommended)
<br>
<br>

- subset()
    + exact search

<br>
<br>
Both of these act against the metadata of the database

## What is metadata?

<br>

- Provided by contributor at the time of data inclusion in the hub database
- Metadata columns:
<font size="3">
    + Title
    + Data provider
    + Data Class
    + Species
    + Taxonomy Id
    + Genome
    + Description
    + Source Type
    + Date Date Added
    + Preparer Classes (Package name / Recipe)
    + Tags
    + Maintainer
    + Data Path
    + Source URL
    + Coordinate 1-based
</font>

## Example: {.build}

<div>
```{r, warning = FALSE}
library(AnnotationHub)
hub = AnnotationHub()
```
</div>
<div>
```{r, warning = FALSE}
hub
```
</div>

## Example

```{r, warning = FALSE}
names(mcols(hub))
```

## Example {.build}

<div>
```{r, warning = FALSE}
length(unique(tolower(hub$species)))
head(unique(tolower(hub$species)))
```
</div>
<div>
```{r, warning = FALSE}
length(unique(hub$rdataclass))
unique(hub$rdataclass)
```
</div>

## Example

```{r, warning = FALSE}
query(hub, "Canis familiaris")
```

## Example

```{r, warning = FALSE}
query(hub, c("Canis familiaris", "GRanges"))
```

## Example {.build}

<div>
```{r, warning = FALSE}
query(hub, c("Canis familiaris", "GRanges", "release-99"))
```
</div>
<div>
Queries are like google searches, the more specific terms, the more specialized
results

</div>

## Example 

Single Bracket "[", Will let you investigate the metadata of a resource WITHOUT downloading 

```{r, warning = FALSE}
hub["AH79003"]
```

## Example {.build}
 
Double Bracket "[[", will download and cache

<div>
```{r, message = TRUE}
dogGR = hub[["AH79003"]]
```
</div>
<div>
```{r}
summary(dogGR)
```
</div>
```{r, message = TRUE}
dogGR = hub[["AH79003"]]
```
<div>

## Example

```{r, warning=FALSE}
library(GenomicFeatures)
TxDb = makeTxDbFromGRanges(dogGR)
TxDb
```

## Example

```{r, warning = FALSE}
query(hub, c("Canis familiaris", "TxDb"))
```

## Example

```{r}
subset(hub, rdataclass=="TxDb" & species =="Canis familiaris")
```

## Example {.build}

<div>
But subset is exact match ...
```{r}
subset(hub, rdataclass=="TxDb" & species =="Canis familiaris" & title=="canFam6")

```
</div>
<div>
Which is why we suggest query
```{r}
query(hub, c("TxDb", "Canis familiaris", "canFam6"))
```
</div>

##

<div style="position:absolute; top:48%; left:33%;">
**Thank you for your attention**
</div>
